core/hotspot/traffic_shaping.go

```diff
@@ -155,6 +159,74 @@ func (c *baseTrafficShapingController) BoundParamIndex() int {
 	return c.paramIndex
 }
 
+// ExtractArgs matches the arg from ctx based on TrafficShapingController
+// return nil if match failed.
+func (c *baseTrafficShapingController) ExtractArgs(ctx *base.EntryContext) []interface{} {
+	if c == nil {
+		return nil
+	}
+	args := make([]interface{}, 0, len(ctx.Input.Args)+len(ctx.Input.Attachments))
```

Comment:
The maxLen is 2,  Maybe we can set len=2.

Reply:
The length of arg and attach is not determined.
Not all of the attach parameters are used as hot spots.
So the length of arg+attach should be more, but the length is measurable.
Why set the length to 2?

Reply:
Currently args only append c.extractArgs(ctx) and c.extractAttachmentArgs(ctx), the maxLen is 2.  len(ctx.Input.Args)+len(ctx.Input.Attachments) may be too big?

Reply:
Remove the slice, paramIndex and paramKey are mutually exclusive to avoid performance loss.

is it ok?

Reply:
ok